Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,664 @@
+        },
+        // We don't display inconclusive on the gutter (user should focus on errors),
+        // We display an error range instead
+        VerificationStatus.Inconclusive => StatusCurrent switch {
+          CurrentStatus.Current => LineVerificationStatus.ErrorRange,
+          CurrentStatus.Obsolete => LineVerificationStatus.ErrorRangeObsolete,
+          CurrentStatus.Verifying => LineVerificationStatus.ErrorRangeVerifying,
+          _ => throw new ArgumentOutOfRangeException()
+        },
+        VerificationStatus.Error => StatusCurrent switch {
+          CurrentStatus.Current => isSingleLine ? LineVerificationStatus.Error : LineVerificationStatus.ErrorRange,
+          CurrentStatus.Obsolete => isSingleLine
+            ? LineVerificationStatus.ErrorObsolete
+            : LineVerificationStatus.ErrorRangeObsolete,
+          CurrentStatus.Verifying => isSingleLine
+            ? LineVerificationStatus.ErrorVerifying
+            : LineVerificationStatus.ErrorRangeVerifying,
+          _ => throw new ArgumentOutOfRangeException()
+        },
+        _ => throw new ArgumentOutOfRangeException()
+      };
+      for (var line = range.Start.Line; line <= range.End.Line; line++) {
+        if (line < 0 || perLineDiagnostics.Length <= line) {
+          // An error occurred? We don't want null pointer exceptions anyway
+          continue;
+        }
+        if ((int)perLineDiagnostics[line] < (int)(targetStatus)) {
+          perLineDiagnostics[line] = targetStatus;
+        }
+      }
+      foreach (var child in Children) {
+        child.RenderInto(perLineDiagnostics,
+          contextHasErrors || StatusVerification == VerificationStatus.Error,
+          contextIsPending ||
+            StatusCurrent == CurrentStatus.Obsolete ||
+          StatusCurrent == CurrentStatus.Verifying,
+          null,
+          Range);
+      }
+      // Ensure that if this is an ImplementationNodeDiagnostic, and children "painted" verified,
+      // and this node is still pending
+      // at least the first line should show pending.
+      if (StatusCurrent == CurrentStatus.Verifying &&
+          perLineDiagnostics.ToList().GetRange(range.Start.Line, range.End.Line - range.Start.Line + 1).All(
+            line => line == LineVerificationStatus.Verified)) {
+        perLineDiagnostics[range.Start.Line] = targetStatus;
+      }
+    }
+
+    // If the verification never starts on this node, it means there is nothing to verify about it.
+    // Returns true if a status was updated
+    public bool SetVerifiedIfPending() {
+      if (StatusCurrent == CurrentStatus.Obsolete) {
+        StatusCurrent = CurrentStatus.Current;
+        StatusVerification = VerificationStatus.Verified;
+        foreach (var child in Children) {
+          child.SetVerifiedIfPending();
+        }
+        return true;
+      }
+
+      return false;
+    }
+
+    public virtual NodeDiagnostic GetCopyForNotification() {
```

Comment:
I solved all my concurrency problems by cloning the node diagnostics before sending them to the client. Indeed, that way the language server protocol can enumerate them while I keep updating the old copy.